packages/flutter/lib/src/widgets/will_pop_scope.dart

```diff
@@ -9,10 +9,66 @@ import 'routes.dart';
 /// Registers a callback to veto attempts by the user to dismiss the enclosing
 /// [ModalRoute].
 ///
+/// {@tool snippet --template=stateful_widget}
+///
+/// Whenever the back button is pressed, you will get a callback at [onWillPop],
+/// which returns a [Future]. If the [Future] returned is true, the screen is
+/// popped and if it is false, the screen is not popped.
+///
+/// ```dart
+/// bool shouldPop = true;
+///
+/// @override
+/// Widget build(BuildContext) {
+///   return WillPopScope (
+///     onWillPop: () async {
+///        return shouldPop;
+///     },
+///     child: Center(),
+///   );
+/// }
+/// ```
+/// {@end-tool}
+///
+/// {@tool dartpad --template=stateful_widget_material}
+///
+/// ```dart preamble
+/// bool shouldPop = true;
+/// ```
+///
+/// ```dart
+/// @override
+/// Widget build(BuildContext context) {
+///   return WillPopScope(
+///       onWillPop: () async {
+///         setState(() {
+///           shouldPop = !shouldPop;
+///         });
+///         return shouldPop;
+///       },
+///       child: Scaffold(
+///           appBar: AppBar(
+///             title: Text("Flutter WillPopScope demp"),
+///             leading: BackButton(),
```

Comment:
I think it might be better to avoid adding the back button manually. You could add a button that pushes a new instance of this widget to have a more real-world navigation stack:

```dart
  bool shouldPop = true;

  @override
  Widget build(BuildContext context) {
    return WillPopScope(
      onWillPop: () async {
        return shouldPop;
      },
      child: Scaffold(
        appBar: AppBar(
          title: Text("Flutter WillPopScope demo"),
        ),
        body: Center(
          child: Column(
            children: [
              OutlinedButton(
                child: Text('Push'),
                onPressed: () {
                  Navigator.of(context)
                      .push(MaterialPageRoute(builder: (context) {
                    return MyStatefulWidget();
                  }));
                },
              ),
              OutlinedButton(
                child: Text('Pop'),
                onPressed: () {
                  setState(() {
                    shouldPop = !shouldPop;
                  });
                },
              ),
              Text("Current value of shouldPop is $shouldPop. "
                  "Tap on back button in appbar to toggle"),
            ],
          ),
        ),
      ),
    );
  }
```

Reply:
Implemented the same